test/cpp/api/module.cpp

```diff
@@ -191,7 +191,7 @@ TEST_CASE("module/clone") {
         buffer = register_buffer("buf", torch::ones({2, 2}));
       }
 
-      Linear l1, l2, l3;
+      Linear l1{nullptr}, l2{nullptr}, l3{nullptr};
```

Comment:
Hmm?

Reply:
So the problem is: `Linear` is like `std::shared_ptr<LinearImpl>`, and we need an empty state for `Linear` (e.g. when you only initialize the module inside `reset()`). I feel like it will be more intuitive to users if `MyModule m;`, where `MyModule` has a default constructor, constructs the `MyModule` with a filled, default constructed value, instead of the usual semantics of `shared_ptr<LinearImpl>`, which would create `MyModule` in the empty state. However, to still allow empty-state construction, I added a constructor that takes `nullptr_t` such that `Linear(nullptr)`. constructs the empty-state `LinearImpl` holder. Does this make sense? Is it intuitive? What do you think?

Reply:
What I am confused about is how there is a "default value" for "Linear".  There's no reasonable default constructible value for Linear, because you have to know what the input and output sizes are. I glanced at the code again and it looks like you're defaulting these parameters to zero. Surely that's unexpected to a user too?

Reply:
You're not defaulting `Linear`, you're defaulting the `shared_ptr<LinearImpl>` to the empty state. It's just like how `at::Tensor()` creates an undefined tensor. The alternative is to disallow an empty state. That becomes annoying in two cases:
1. Late initialization, e.g. the fact that many of our modules do most of their initialization work inside `reset()`
2. Containers, imagine someone wants to store `std::vector<Linear>`.
I feel like people would work around this by storing `std::shared_ptr<Linear>`, but that would defeat the purpose of our whole pImpl machinery. What do you think?